### DISCUSSION TITLE
gatherer package_version returns only first package if multiple versions are installed

### DISCUSSION THREAD

**Discussion by scmschmidt** [trento-project/wanda#150] [Q&A] [answered]
During test development I managed to install two versions of the same package:

```
# rpm -q sbd
sbd-1.4.2+20210305.926b554-3.3.1.x86_64
sbd-1.3.1+20180507.e102d9e-1.6.x86_64
```

The gatherer only returns the package with the highest version: 

```
# trento-agent facts gather --gatherer package_version --argument sbd
INFO[2023-01-03 13:31:04] Using config file: /etc/trento/agent.yaml    
INFO[2023-01-03 13:31:04] loading plugins                              
INFO[2023-01-03 13:31:04] Starting package_version facts gathering process 
INFO[2023-01-03 13:31:04] Requested package_version facts gathered     
INFO[2023-01-03 13:31:04] Gathered fact for "package_version" with argument "sbd": 
INFO[2023-01-03 13:31:04] {
  "Name": "sbd",
  "CheckID": "",
  "Value": {
    "Value": "1.4.2+20210305.926b5541.3.1+20180507.e102d9e"
  },
  "Error": null
} 
```

The question is, if the gatherer should not take this in consideration and should return a list?
In my case this was a mistake, but this also prevents Trento from detecting such an error (we should discuss, if this is wanted at all), but also in the future it might me valid to have multiple packages installed.

**Comment by nelsonkopliku** [comment]
Summarizing a slack thread about this topic, so we can bring back the discussion here.

Multiple installed versions topic is narrowed down to kernels only, so we might end up implementing an `uname` gatherer to check the kernel version.
Package version gatherer remains untouched.

**Comment by scmschmidt** [comment]
We discussed the topic in our Trento team meeting and @abravosuse opened a Jira issue to track the progress. Neverthe less the technical discussion can be done here.

With the latest version trento-agent-1.2.0+git.dev41.1674042848.d23f6de-lp154.1.1.x86_64 the behavior has already changed. If asked for a package, it returns all versions in one string without separator (example for two installed sbd packages: 1.5.0+20210720.f4ca41f1.3.1+20180507.e102d9e). The ordering seems to be highest version first.
In such a case asking for a specific package version (of the installed ones), the gatherer always returns  -1)

When asking for a package, can we get a list instead? At least a separator is required.
Also we think, that additional information about a package is required. The package release can be a distinguishing factor as well to check the origin for instance. In short all the information of `rpm -qi `  or `zypper info` would be helpful.

As you use `zypper` already, please make sure that the used configuration and passed arguments result in a short-running, non-blocking and resource-friendly process. 

The Jira issue has some additional ideas, we can discuss in the future which would help to write more elaborate checks (like the output of `rpm -ql,` maybe even with checksum or the result of `rpm -V`. But with the last one we'll be in the resource discussion again, since this can be IO intensive.

**Reply by nelsonkopliku** [reply]
Resuming this.

I managed to install two different versions of the `sdb` package, as in the example (`1.5.1+20221128.8ec8e01` and `1.5.1+20211116.6bb085f`), however I found a bit hacky getting there, though, because I found myself editing `/etc/zypp/zypp.conf` to allow `sbd` in multiversion and accepting file conflicts when installing the _other_ version in order to get two installed.

As you also mentioned that was a mistake in your case, so I guess this a scenario where end users would not be (often)

---

However, even though an edge case, I get the opportunity to be able to detect the scenario.

Returning a list would not be a big deal, as adjusting the queryformat to `rpm -q --qf` would do the trick and we could also extract other information if needed
```bash
$ rpm -q --qf 'VERSION=%{VERSION} INSTALL_TIME=%{INSTALLTIME:date} RELEASE=%{RELEASE} VENDOR="%{VENDOR}"\n' sbd
VERSION=1.5.1+20221128.8ec8e01 INSTALL_TIME=Wed Feb  8 14:09:41 2023 RELEASE=150400.3.3.1 VENDOR="SUSE LLC <https://www.suse.com/>"
VERSION=1.5.1+20211116.6bb085f INSTALL_TIME=Wed Feb  8 14:12:59 2023 RELEASE=150400.1.5 VENDOR="SUSE LLC <https://www.suse.com/>"
```
However the point is that since what we get is _a list of installed versions for a package `<package-name>`_, we have to decide which one is the one we want to match against in the check.
Options are:
- the last installed (might not be the highest version available)
- the highest version (might not be the last installed version)

with this information we can improve the gatherer and make sure to still return just a string of the detected version (based on the defined rule). This keeps it simple also for comparison scenario.

In the future we can expose more information and evolve the thing.

@scmschmidt @angelabriel what do you think?

**Reply by angelabriel** [reply]
From my point of view we should use the **last installed** version for the comparison (for now) and should mention this in the gatherer documentation so that it will be clear for all users.

For the future it will be really helpful to get more information (as you already mentioned).
And may be we find a way to determine which is the **used** version (latest installed version and used version will mostly be the same, but for kernel package and some other packages using '**update-alternatives**' it can be different). And then we can change or better enhance the comparison option.

**Reply by nelsonkopliku** [reply]
Alright, last installed one, even if it is not the highest version.

Plus one on documenting it :+1: and yes determining the **used** version is another topic indeed.
`kernel` might have its own gatherer and other packages could pass through **update-alternatives**.

Sounds like a plan :wink:

**Reply by abravosuse** [reply]
I support Angela's suggestion

**Comment by scmschmidt** [comment]
> As you also mentioned that was a mistake in your case, so I guess this a scenario where end users would not be (often)
Agreed.  A sloppy install/update script or an dauntless admin could causing it.

I also agree, that the last installed version is the best result. It brought the files, the system is going to use.

Just to get it right. The gatherer will still return a string. The idea to get a list of objects (containing the details ob a package) will come later?
If the answer is yes, will the gatherer return an additional entry for that list? 
I just want to make sure, that with future enhancements we do not introduce incompatibilities --> check rewrites.
Let us define that now as best as we can.

**Reply by nelsonkopliku** [reply]
Totally fair point. In order to reduce checks rewrite we can leverage a map/object.

**Option 1**: We return a single object, the latest installed version of the `package`
```rs
#{
    name: "package-name"
    version: "the-version"
}   
```
**Option 2**: we return a list of of previous objects

Both of them open for extension of information provided.

Option 1 sounds enough for what we do. What do you think about it? 
~What scenarios where we query for many packages at once from this gatherer am I missing? Option 2 would fit better here.~

EDIT: What scenarios where we use the different versions am I missing? Option 2 would fit better here.

**Comment by scmschmidt** [comment]
I vote for Option 2.  I know, in almost all cases this list will have one entry, but as we know, there can be exceptions.

With our first approach, having the latest installed version only, it would have only one element:
```
[ 
  {
    name: "package-name"
    version: "the-version"
  } 
]
```

Later (or even now?) we can have all versions and the sorting makes clear which came latest (maybe the first element?):

```
[ 
  {
    // I'm the latest version
    name: "package-name"
    version: "the-version"
  }
  {
   // I'm another version, but was installed earlier
    name: "package-name"
    version: "the-version"
  }
 ]
```

Don't know if it is possible or even makes sense to have the list sorted in installation order.

Later we can add additional attributes, like vendor, etc.

What do you and especially our dearest check writer (@angelabriel) thinks? ;-)

**Reply by nelsonkopliku** [reply]
I am fine in returning the whole list of detected versions.

We'll just need to access first element of that list in the checks spec either via `package_fact[0].version` or `package_fact.first().version`.

Sounds good.

**Reply by angelabriel** [reply]
yes, fine for me too.
To summarize my understanding:
comparison (argument: package,version) is done against the latest installed version of the package and returns -1,0,1
ask for version (argument: package) returns a list ordered in reverse installation order (last installed on top of the list)
and when used with 'expect_same: ' I can simply refer to` facts.package_fact.first().version`

**Reply by nelsonkopliku** [reply]
Yep. Version comparison will internally use the latest installed version.

**Comment by nelsonkopliku** [comment, marked as answer]
@angelabriel @scmschmidt here we go
Agent changes -> https://github.com/trento-project/agent/pull/18
Updated doc ->https://github.com/trento-project/wanda/pull/182
